src/analysis/souffle/taint.dl

```diff
@@ -35,28 +40,34 @@
 // they have Tags).
 .decl mayHaveTag(accessPath: AccessPath, tag: Tag)
 
-// There is a claim that `accessPath` definitely has `tag` taint.
-.decl claimHasTag(accessPath: AccessPath, tag: Tag)
+// There is a claim that `accessPath` definitely has `tag` taint. This claim
+// is made by `principal`.
+.decl claimHasTag(principal: Principal, accessPath: AccessPath, tag: Tag)
+
+// Any Principal used in a claim is in the universe of isPrincipal.
+isPrincipal(principal) :- claimHasTag(principal, _, _).
 
 //-----------------------------------------------------------------------------
 // Rules
 //-----------------------------------------------------------------------------
-mayHaveTag(tgt, tag) :- claimHasTag(tgt, tag).
+hasTag(tgt, tag) :- claimHasTag(principal, tgt, tag), ownsTag(principal, tag).
+
+mayHaveTag(tgt, tag) :- hasTag(tgt, tag).
+
 mayHaveTag(tgt, tag) :-
-    edge(src, tgt), mayHaveTag(src, tag),
-    !downgrades(tgt, tag).
+  edge(src, tgt), mayHaveTag(src, tag), !downgrades(tgt, tag).
 
 //-----------------------------------------------------------------------------
 // Universe Populating Rules
 //-----------------------------------------------------------------------------
 // These rules add objects to the set of things that exist (i.e. relations like
 // `isObject(..)`) whenever they are mentioned in other rules.
-isAccessPath(path) :- claimHasTag(path, _).
+isAccessPath(path) :- claimHasTag(_, path, _).
+isAccessPath(path) :- hasTag(path, _).
 isAccessPath(path) :- mayHaveTag(path, _).
 
 // Symbols used in hasTag or mayHaveTag are tags
-isTag(tag) :- claimHasTag(_, tag).
-isTag(tag) :- mayHaveTag(_, tag).
+isTag(tag) :- claimHasTag(_, _, tag).
```

Comment:
Leaving it there would cause a circularity with `ownsTag` defined in the "all principals own all tags" fashion. The reason is that `mayHaveTag` is a relation consisting of inferred facts, many of which use `isTag` to ground their variables.

I'm thinking that, in general, it's a bad idea to have these universe-inferring rules depend upon anything other than a relation which will be populated with base facts. It opens the door for this kind of circularity. This was the only one that needed to be removed to fix the problem, but any rule populated from `mayHaveTag` could cause it. I think we may need to do a cleanup pass later.

Reply:
Filed as #126

Reply:
I agree with not having universe-inferring rules. That is what I realized when looking at some other part of the PR.